splink/expectation_maximisation.py

```diff
@@ -56,6 +51,26 @@ def compute_new_parameters_sql(settings_obj: Settings):
     return sql
 
 
+def calculate_m_and_u_probability_averages(m_u_df):
+
+    data = m_u_df.copy()
```

Comment:
I think (as you said in your comment) it's better to write this as sql.  Particularly since the pandas code is quite fiddly here so it's not obvious what's going on.  here's a suggested implementation which I think is a bit more 'self documenting'.  It'd need updating to reflect the `m_count` an `u_count` column above (which would help further with 'self documentation;).

```
    con = duckdb.connect(":memory:")
    con.register("m_u_df", m_u_df)

    sql = """
    select
        comparison_vector_value,
        output_column_name,
        m_probability/sum(m_probability) OVER (PARTITION BY output_column_name)
            AS m_probability,
        u_probability/sum(u_probability) OVER (PARTITION BY output_column_name)
            AS u_probability
    from m_u_df
    where comparison_vector_value != -1
    and output_column_name != '_probability_two_random_records_match'

    union all

    select
        comparison_vector_value,
        output_column_name,
        m_probability,
        u_probability
    from m_u_df
    where output_column_name = '_probability_two_random_records_match'
    order by output_column_name, comparison_vector_value asc
    """

    return con.query(sql).to_df().to_dict("records")
```